### ISSUE TITLE
Signed releases with Go 1.5

### ISSUE THREAD

**Issue by jbenet** [jbenet/go-ipfs#957]
Once Go 1.5 lands we can compile all archs in one place. We can then do signed releases.
It would be ideal to deploy these releases using ipfs itself, using the gateways.

**Comment by sebie** [comment]
+1 On this. And for now, I would say a md5 hash wouldn't be bad.

**Comment by gwillen** [comment]
+1 for sure. I see ipfs going a long way; signed releases are an obvious step and would fit in pretty well with a distribution model that works via ipfs itself.

**Comment by jbenet** [comment]
@gwillen absolutely. I'm waiting on Go 1.5 so that I can cross compile locally.

Actually, @okket just got it all working based on the syncthing build scripts. See https://gist.github.com/okket/60f2897227fe59dedf76 -- happy to start compiling in a local machine and start doing signed releases.

**Comment by jbenet** [comment]
@gwillen absolutely. I was waiting on Go 1.5 so that I could cross compile locally. But actually, @okket just got it all working based on the syncthing build scripts. See https://gist.github.com/okket/60f2897227fe59dedf76 -- happy to start compiling in a local machine and start doing signed releases. Parts still needed:

- [x] get cross compiling working (@okket)
- [ ] code-signing script based on best practices
- [ ] new build targets in a makefile
- [ ] tool for users to check our code-signing
- [ ] fix pinning bugs (#1051 #1008)
- [ ] S3 pinning

- @Luzifer, happy to incorporate gobuilder in any of this, though the compilation probably will have to happen locally on my machine, at least for official releases. 
- With deterministic builds, we could also have @whyrusleeping and other community members do their own builds to verify my machine/I haven't been compromised (and gobuilder)
- would be nice to see how auto-updating can figure into this. @inconshreveable's equinox.io is a great solution in general, though it's not self-hosting (i.e. using ipfs) and i'd prefer it if we dog-fooded our own stuff for deployments. (maybe there's a way to do that?)

**Comment by okket** [comment]
How to cross compile ipfs locally:

First, install ```goxc``` with ```go get github.com/laher/goxc```, then install the cross compile toolchain binaries with ```goxc -t```. This takes a long time.

Now cross build:
```
goxc -q -bc='linux,!arm darwin' -d=~/downloads -pv=0.2.4-alpha1 -tasks-=validate -build-gcflags='-i -wd cmd/ipfs
```

For Windows binaries your need two small tweaks: Install a dependency with ```go get github.com/olekukonko/ts``` and the cross compile command needs to disable FUSE:

```
goxc -q -v -bc='windows,386' -d=~/downloads -pv=0.2.4-alpha1 -tasks-=validate -build-gcflags='-i' -wd cmd/ipfs -build-tags=nofuse
```

**Comment by okket** [comment]
Signing binaries on OS X is really easy:

```
codesign -s "<sign identity substring>" ipfs
```

Check:

```
$ codesign -v -d ipfs 
Executable=/Users/okketimm/Downloads/ipfs_0.2.4-alpha1_darwin_amd64/ipfs
Identifier=ipfs
Format=Mach-O thin (x86_64)
CodeDirectory v=20200 size=98008 flags=0x0(none) hashes=4895+2 location=embedded
Signature size=4343
Signed Time=13.04.2015 08:37:03
Info.plist=not bound
TeamIdentifier=H72RJK679G
Sealed Resources=none
Internal requirements count=1 size=180
```

**Comment by whyrusleeping** [comment]
with go 1.5, cross compilation is:
```
GOOS=windows go build
GOARCH=arm go build
```

**Comment by vardars** [comment]
+1 on this
+1 on that (md5)

**Comment by ErikBjare** [comment]
+1, both signed and reproducible releases would be awesome.

**Comment by aakilfernandes** [comment]
+1

**Comment by peterhil** [comment]
+1

**Comment by stylesuxx** [comment]
+1

**Comment by rugk** [comment]
:+1:

**Comment by Kubuxu** [comment]
also
:+1: for signing
:-1: for md5

**Comment by dandart** [comment]
I'd recommend SHA512 to make it more futureproof

**Comment by jbenet** [comment]
@dandart 
- https://github.com/jbenet/multihash 
- also, sha512 does not confer significant algorithmic advantage over sha256-- it's the same algorithm with different initial parameters + rounds, so an algorithmic break is liable to break them both.
- we'll include several digests

**Comment by cjcase** [comment]
yay to signing, I'd go with PGP (GnuPG)

**Comment by johncant** [comment]
Yes please!

**Comment by j4nu5** [comment]
+1

**Comment by faebser** [comment]
+1

**Comment by zciendor** [comment]
You could at least publish signed SHASUMS with each release for the prebuilt binaries. This is independent from Go.
```
sha256sum <BUILD-OUTPUT-DIR>/* | gpg2 --clearsign --armor --output SHA256SUMS.asc
```
This will create an `SHA256SUMS.asc` file (like Bitcoin Core ships with each release) containing all the hashsums of the various binaries and put it in the same directory as the release (e.g. https://ipfs.io/ipns/dist.ipfs.io/go-ipfs/v0.4.15).

All you have to do is publish the corresponding `public-key` on various locations/sources (Github, ipfs.io website, IRC, forums, ...)
